File: to_be_solved/arrays/121-best-time-to-buy-and-sell-stock.ts
// LeetCode 121: Best Time to Buy and Sell Stock
// You are given an array prices where prices[i] is the price of a given stock on the ith day.
// You want to maximize your profit by choosing a single day to buy one stock and choosing a different day in the future to sell that stock.

function maxProfit(prices: number[]): number {
    let minPrice = prices[0];
    let maxProfit = 0;
    
    for (let i = 1; i < prices.length; i++) {
        if (prices[i] < minPrice) {
            minPrice = prices[i];
        } else {
            maxProfit = Math.max(maxProfit, prices[i] - minPrice);
        }
    }
    
    return maxProfit;
}

// Test cases
console.log(maxProfit([7, 1, 5, 3, 6, 4])); // 5 (buy at 1, sell at 6)
console.log(maxProfit([7, 6, 4, 3, 1])); // 0 (no profit possible)
console.log(maxProfit([1, 2, 3, 4, 5])); // 4 (buy at 1, sell at 5)

File: src_py/121-best-time-to-buy-and-sell-stock.py
"""
121. Best Time to Buy and Sell Stock
Difficulty: Easy

You are given an array prices where prices[i] is the price of a given stock on
the i-th day. Maximize your profit by choosing a single day to buy and a
different day in the future to sell. Return the maximum profit, or 0.
"""

from typing import List


def max_profit(prices: List[int]) -> int:
    if not prices:
        return 0
    min_price = prices[0]
    best = 0
    for price in prices[1:]:
        best = max(best, price - min_price)
        min_price = min(min_price, price)
    return best


if __name__ == "__main__":
    assert max_profit([7, 1, 5, 3, 6, 4]) == 5
    assert max_profit([7, 6, 4, 3, 1]) == 0
    assert max_profit([]) == 0
    print("121 OK:", max_profit([7, 1, 5, 3, 6, 4]))
